homeassistant/components/ring/__init__.py

```diff
@@ -104,42 +103,46 @@ async def async_remove_config_entry_device(
     return True
 
 
-async def _migrate_old_unique_ids(hass: HomeAssistant, entry_id: str) -> None:
-    entity_registry = er.async_get(hass)
-
-    @callback
-    def _async_migrator(entity_entry: er.RegistryEntry) -> dict[str, str] | None:
-        # Old format for camera and light was int
-        unique_id = cast(str | int, entity_entry.unique_id)
-        if isinstance(unique_id, int):
-            new_unique_id = str(unique_id)
-            if existing_entity_id := entity_registry.async_get_entity_id(
-                entity_entry.domain, entity_entry.platform, new_unique_id
-            ):
-                _LOGGER.error(
-                    "Cannot migrate to unique_id '%s', already exists for '%s', "
-                    "You may have to delete unavailable ring entities",
-                    new_unique_id,
-                    existing_entity_id,
-                )
-                return None
-            _LOGGER.debug("Fixing non string unique id %s", entity_entry.unique_id)
-            return {"new_unique_id": new_unique_id}
-        return None
-
-    await er.async_migrate_entries(hass, entry_id, _async_migrator)
-
-
 async def async_migrate_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Migrate old config entry."""
     entry_version = entry.version
     entry_minor_version = entry.minor_version
+    entry_id = entry.entry_id
 
     new_minor_version = 2
     if entry_version == 1 and entry_minor_version == 1:
         _LOGGER.debug(
             "Migrating from version %s.%s", entry_version, entry_minor_version
         )
+        # Migrate non-str unique ids
+        # This step used to run unconditionally from async_setup_entry
+        entity_registry = er.async_get(hass)
+
+        @callback
+        def _async_str_unique_id_migrator(
+            entity_entry: er.RegistryEntry,
+        ) -> dict[str, str] | None:
+            # Old format for camera and light was int
+            unique_id = cast(str | int, entity_entry.unique_id)
+            if isinstance(unique_id, int):
+                new_unique_id = str(unique_id)
+                if existing_entity_id := entity_registry.async_get_entity_id(
+                    entity_entry.domain, entity_entry.platform, new_unique_id
+                ):
+                    _LOGGER.error(
+                        "Cannot migrate to unique_id '%s', already exists for '%s', "
+                        "You may have to delete unavailable ring entities",
+                        new_unique_id,
+                        existing_entity_id,
+                    )
+                    return None
+                _LOGGER.debug("Fixing non string unique id %s", entity_entry.unique_id)
+                return {"new_unique_id": new_unique_id}
+            return None
+
+        await er.async_migrate_entries(hass, entry_id, _async_str_unique_id_migrator)
```

Comment:
Are we sure that all config entries after version 1.1 are correct?

Reply:
Yes, the change to convert the int unique ids went in a long time before the migration step from 1.1 > 1.2 went in